src/py21cmfast/src/HaloBox.c

```diff
@@ -329,32 +477,219 @@ int get_box_averages(double redshift, double norm_esc, double alpha_esc, double
+    if(user_params_stoc->AVG_BELOW_SAMPLER && M_min < user_params_stoc->SAMPLER_MIN_MASS){
+        LOG_DEBUG("SUB-SAMPLER",consts->redshift);
+        LOG_DEBUG("Exp. averages: (HM %11.3e, NION %11.3e, SFR %11.3e, SFR_MINI %11.3e)",averages_sub_expected.halo_mass,averages_sub_expected.n_ion,
+                                                                                                averages_sub_expected.halo_sfr,averages_sub_expected.sfr_mini);
+        LOG_DEBUG("Box  averages: (HM %11.3e, NION %11.3e, SFR %11.3e, SFR_MINI %11.3e)",averages_subsampler->halo_mass,averages_subsampler->n_ion,
+                                                                                                averages_subsampler->halo_sfr,averages_subsampler->sfr_mini);
+    }
+    LOG_DEBUG("Turnovers: ACG %11.3e global %11.3e",averages_box->m_turn_a,consts->mturn_a_nofb);
+    LOG_DEBUG("MCG  %11.3e nofb %11.3e",averages_box->m_turn_a,consts->mturn_m_nofb);
+    LOG_DEBUG("Reion %11.3e",averages_box->m_turn_r);
+}
+
+void sum_halos_onto_grid(struct InitialConditions *ini_boxes, struct TsBox *previous_spin_temp, struct IonizedBox * previous_ionize_box,
+                          struct PerturbHaloField *halos, struct HaloBoxConstants *consts, struct HaloBox *grids, struct HaloProperties *averages){
+    double redshift = consts->redshift;
+    //averages
+    double hm_avg=0.,sm_avg=0.,sfr_avg=0.;
+    double sm_avg_mini=0.,sfr_avg_mini=0.;
+    double M_turn_a_avg=0.,M_turn_m_avg=0.,M_turn_r_avg=0.;
+    double n_ion_avg=0., wsfr_avg=0., xray_avg=0.;
+    //counts
+    unsigned long long int total_n_halos, n_halos_cut=0.;
+
+    double cell_volume = VOLUME / HII_TOT_NUM_PIXELS;
+    #pragma omp parallel num_threads(user_params_stoc->N_THREADS)
+    {
+        int x,y,z;
+        unsigned long long int i_halo,i_cell;
+        double m,nion,sfr,wsfr,sfr_mini,stars_mini,stars,xray;
+        double J21_val, Gamma12_val, zre_val;
+
+        double curr_vcb = consts->vcb_norel;
+        double M_turn_m = consts->mturn_m_nofb;
+        double M_turn_a = consts->mturn_a_nofb;
+        double M_turn_r;
+
+        float in_props[3];
+        float out_props[7];
+
+    #pragma omp for reduction(+:hm_avg,sm_avg,sm_avg_mini,sfr_avg,sfr_avg_mini,M_turn_a_avg,M_turn_m_avg,M_turn_r_avg,n_halos_cut)
+        for(i_halo=0; i_halo<halos->n_halos; i_halo++){
+            m = halos->halo_masses[i_halo];
+            //It is sometimes useful to make cuts to the halo catalogues before gridding.
+            //  We implement this in a simple way, if the user sets a halo's mass to zero we skip it
+            if(m == 0.){
+                n_halos_cut++;
+                continue;
+            }
+            x = halos->halo_coords[0+3*i_halo]; //NOTE:PerturbHaloField is on HII_DIM, HaloField is on DIM
+            y = halos->halo_coords[1+3*i_halo];
+            z = halos->halo_coords[2+3*i_halo];
```

Comment:
are you hard-coding `DIM = 3*HII_DIM` here?

Reply:
Nope, the `halo_coords` field is laid out (x0,y0,z0,x1,y1,z1...) the 3 is for the 3 dimensions

Reply:
thanks!